### PR TITLE
Added support for throwing if name not found to ResourceManagerStringLocalizer

### PR DESCRIPTION
**Pull request by leonardlabat**
- Added the ResourceLookupBehavior enum
- Using it into the LocalizationOptions
- Added it to the ResourceMangerStringLocalizer
- Refactored all the automatic fallbacks to the resource name (now use a
  single method)
- Added unit tests for both specific behaviors

Addresses #13

### CODE REVIEW

**Comment by hishamco** [comment]
If I'm not wrong, this is in the backlog, your thought @DamianEdwards 
/cc @kichalla @Eilon 

thanks @leonardlabat

**Comment by Eilon** [comment]
@DamianEdwards thoughts on this?

**Comment by DamianEdwards** [comment]
Too late for RC2/RTM but we can consider this for later.

File: global.json [hunk 1]
```diff
@@ -1,3 +1,6 @@
 {
-  "projects": [ "src" ]
+  "projects": [ "src" ],
+  "sdk": {
```
Comment (DamianEdwards) [inline global.json]: Remove the "sdk" node completely

File: src/Microsoft.Extensions.Localization/LocalizationOptions.cs [hunk 2]
```diff
@@ -12,5 +12,10 @@ public class LocalizationOptions
         /// The relative path under application root where resource files are located.
         /// </summary>
         public string ResourcesPath { get; set; } = string.Empty;
+
+        /// <summary>
+        /// The behavior to apply when a given resource key appears to be missing
```
Comment (DamianEdwards) [inline src/Microsoft.Extensions.Localization/LocalizationOptions.cs]: "The lookup behavior to use when a given resource key is not found."

File: src/Microsoft.Extensions.Localization/ResourceLookupBehavior.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved. 
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+namespace Microsoft.Extensions.Localization
+{
+    /// <summary>
+    /// Behaviors to be used when a resource key appears to be missing
+    /// </summary>
+    public enum ResourceLookupBehavior
```
Comment (DamianEdwards) [inline src/Microsoft.Extensions.Localization/ResourceLookupBehavior.cs]: "Defines the lookup behavior to use when a given resource key is not found."

File: src/Microsoft.Extensions.Localization/ResourceLookupBehavior.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        /// <summary>
+        /// Use the resource key as a fallback
```
Comment (DamianEdwards) [inline src/Microsoft.Extensions.Localization/ResourceLookupBehavior.cs]: "Use the resource key as the string if a localized string is not found."

File: src/Microsoft.Extensions.Localization/ResourceLookupBehavior.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        /// </summary>
+        UseNameIfNotFound,
+        /// <summary>
```
Comment (DamianEdwards) [inline src/Microsoft.Extensions.Localization/ResourceLookupBehavior.cs]: new line

File: src/Microsoft.Extensions.Localization/ResourceLookupBehavior.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        /// An exception is thrown to the caller
```
Comment (DamianEdwards) [inline src/Microsoft.Extensions.Localization/ResourceLookupBehavior.cs]: "Throw an exception if a localized string is not found."

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 7]
```diff
@@ -157,7 +163,26 @@ protected IEnumerable<LocalizedString> GetAllStrings(bool includeAncestorCulture
...
             {
                 var value = GetStringSafely(name, culture);
-                yield return new LocalizedString(name, value ?? name, resourceNotFound: value == null);
+                yield return new LocalizedString(name, value, resourceNotFound: value == null);
+            }
+        }
+
+        private string ApplyResourceLookupBehavior(string name, string value)
+        {
+            switch(_resourceLookupBehavior)
```
Comment (DamianEdwards) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: Just use an `if` block here

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 8]
```diff
@@ -157,7 +163,26 @@ protected IEnumerable<LocalizedString> GetAllStrings(bool includeAncestorCulture
...
+            {
+                case ResourceLookupBehavior.ThrowIfNotFound:
+                    if(value == null)
```
Comment (DamianEdwards) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: formatting

File: src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs [hunk 9]
```diff
@@ -157,7 +163,26 @@ protected IEnumerable<LocalizedString> GetAllStrings(bool includeAncestorCulture
...
+                    {
+                        throw new ArgumentOutOfRangeException(
+                            nameof(name),
+                            name,
+                            "The name of the string resource is missing from the resource source.");
```
Comment (DamianEdwards) [inline src/Microsoft.Extensions.Localization/ResourceManagerStringLocalizer.cs]: Doesn't seem like the right exception to throw. Probably should throw a new custom exception, e.g. `LocalizedStringNotFoundException`. The error message will need to be localized with traditional resources, e.g. https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Resources.resx

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs [hunk 10]
```diff
@@ -8,6 +8,7 @@
 using System.Resources;
 using Microsoft.Extensions.Localization.Internal;
 using Xunit;
+using System;
```
Comment (DamianEdwards) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs]: Sort usings

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs [hunk 11]
```diff
@@ -59,6 +60,37 @@ public void EnumeratorCacheIsScopedByAssembly()
             Assert.Equal(expectedCallCount, resourceAssembly2.GetManifestResourceStreamCallCount);
         }
 
+        [Fact]
+        public void ResourceLookupBehaviorUseNameIfNotFound()
```
Comment (DamianEdwards) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs]: The name should be descriptive of the assertion as well, e.g. `ReturnsNameForMissingStringWhenResourceLookupBehaviorIsUseNameIfNotFound`

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs [hunk 12]
```diff
@@ -59,6 +60,37 @@ public void EnumeratorCacheIsScopedByAssembly()
...
+        public void ResourceLookupBehaviorThrowIfNotFound()
+        {
+            // Arrange
+            var resourceNamesCache = new ResourceNamesCache();
+            var baseName = "test";
+            var resourceAssembly = new TestAssemblyWrapper("Assembly1");
+            var resourceManager = new TestResourceManager(baseName, resourceAssembly.Assembly);
+            var localizer = new ResourceManagerStringLocalizer(resourceManager, resourceAssembly, baseName, resourceNamesCache, ResourceLookupBehavior.ThrowIfNotFound);
+
+            // Ac & Assert
```
Comment (DamianEdwards) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs]: typo

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs [hunk 13]
```diff
@@ -59,6 +60,37 @@ public void EnumeratorCacheIsScopedByAssembly()
...
+
+            // Act
+            var value = localizer["test"];
+
+            // Assert
+            Assert.Equal("test", value);
+        }
+
+        [Fact]
+        public void ResourceLookupBehaviorThrowIfNotFound()
```
Comment (DamianEdwards) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs]: The name should be descriptive of the assertion as well, e.g. `ThrowsForMissingStringWhenResourceLookupBehaviorIsThrowIfNotFound`

**Comment by Eilon** [comment]
Moving the PR to the backlog; we will revisit in the next release.

**Comment by hishamco** [comment]
@Eilon this should be closed too, after you closed issue #13

**Pull request closed without merging**